lib/src/merge.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use itertools::Itertools;
+
+/// Attempt to resolve trivial conflicts between the inputs. Even indexes are
+/// positive terms, and odd indexes are negative terms. The length must be odd.
+pub fn trivial_merge<T>(input: &[T]) -> Option<T>
+where
+    T: Eq + Hash + Clone,
+{
+    assert_eq!(input.len() % 2, 1);
+
+    // Check if all sides made the same change.
+    // This matches what Git and Mercurial do (in the 3-way case at least), but not
+    // what Darcs and Pijul do. It means that repeated 3-way merging of multiple
+    // trees may give different results depending on the order of merging.
+    let mut all_changes_same = true;
+    for i in 2..input.len() {
+        if input[i] != input[i - 2] {
+            all_changes_same = false;
+            break;
+        }
+    }
+    if all_changes_same {
+        return Some(input[0].clone());
+    }
+
+    // Number of occurrences of each value, with positive indexes counted as +1 and
+    // negative as -1.
+    let mut counts: HashMap<T, i32> = HashMap::new();
+    for (i, value) in input.iter().enumerate() {
+        let delta = if i % 2 == 0 { 1 } else { -1 };
+        counts
+            .entry(value.clone())
+            .and_modify(|e| *e += delta)
+            .or_insert(delta);
+    }
+
+    let x = counts
+        .iter()
+        .filter_map(|(k, v)| if *v != 0 { Some((k, *v)) } else { None })
+        .collect_vec();
+    match x[..] {
+        [(k, 1)] => Some(k.clone()),
+        _ => None,
+    }
```

Comment:
Checking my understanding: `counts` is a hashmap of each term to the number of times it appears (after adding and subtracting). We return `Some()` (trivial merge succeeds) iff there is exactly one term and it appears once.

Reply:
Yep, correct. I've added some more documentation.